tokenizers/src/pre_tokenizers/split.rs

```diff
@@ -0,0 +1,145 @@
+use std::fmt;
+
+use regex::Regex;
+use serde::{Deserialize, Deserializer, Serialize};
+
+// use crate::pattern::Pattern;
+use crate::tokenizer::{
+    pattern::{Invert, Pattern}, PreTokenizedString, PreTokenizer, Result, SplitDelimiterBehavior,
+};
+use serde::de::{Error, Visitor};
+
+#[derive(Clone, Debug, Serialize)]
+#[serde(tag = "type")]
+pub struct Split {
+    #[serde(default = "default_regex", skip)]
+    re: Regex,
+    #[serde(default = "default_behavior", skip)]
+    behavior: SplitDelimiterBehavior,
+    invert: bool
+}
+
+fn default_regex() -> Regex {
+    Regex::new(r"\w+|[^\w\s]+").unwrap()
+}
+
+fn default_behavior() -> SplitDelimiterBehavior {
+    SplitDelimiterBehavior::Removed
+}
+
+impl Default for Split {
+    fn default() -> Self {
+        Self {
+            re: default_regex(),
+            behavior: default_behavior(),
+            invert: false,
+        }
+    }
+}
+
+impl Split {
+    pub fn new(re: Regex, behavior: SplitDelimiterBehavior, invert: bool) -> Self {
+        Split {
+            re,
+            behavior,
+            invert
+        }
+    }
+}
+
+impl PreTokenizer for Split {
+    fn pre_tokenize(&self, pretokenized: &mut PreTokenizedString) -> Result<()> {
+        use SplitDelimiterBehavior::*;
```

Comment:
Usually imports are at the beginning of the file. It should always be at the top level module
Exception is tests within a file (where imports are still at the beginning of the module and not in functions)
Tests is usually a module at the end of a file.

Reply:
That's a good way to avoid polluting the global scope with enum variants though, and something you'll see very often in Rust. There are multiple places in the lib where we use it.